### PR TITLE
API server discovery.

### PR DESCRIPTION
**Pull request by jmalloc**
This PR adds a discovery system for the config API.

The package API is centered around the `TargetObserver` and `ClientObserver` interfaces.

`TargetObserver` implementations are notified when a target is available. This means that some discovery mechanism has identified a server that MAY implement the config API.

`ClientObserver` implementations are notified when a target can be successfully dialed, and actually does implement the config API. The `Connector` is a `TargetObserver` responsible for making these connections and checking if the config API is supported before notifying some other `ClientObserver`.

Lastly, this PR includes a very basic DNS-based discovery mechanism. In the future I would like to add conventional DNS-SD, Bonjour and Docker-based mechanisms.

### CODE REVIEW

File: api/discovery/connector.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+	conn, err := dial(ctx, w.Target)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	return w.watch(ctx, conn)
+}
+
+// watch checks if conn supports the config API and notifies obs accordingly.
```
Comment (danilvpetrov) [inline api/discovery/connector.go]:
```suggestion
// watch checks if conn supports the config API and notifies observers accordingly.
```

File: api/discovery/connector.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+//
+// It blocks until ctx is canceled or the connection is severed, at which point
+// the obs is notified of the disconnection.
```
Comment (danilvpetrov) [inline api/discovery/connector.go]:
```suggestion
// watch blocks until ctx is canceled or the connection is severed, at which point
// the observer is notified of the disconnection.
```

File: api/discovery/connector_test.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+		})
+
+		It("does not notify the observer", func() {
+			connector.TargetAvailable(target)
+			defer connector.TargetUnavailable(target)
+			<-ctx.Done() // wait out the timeout
+		})
+	})
+
+	Context("when the server is does not implement the config API", func() {
```
Comment (danilvpetrov) [inline api/discovery/connector_test.go]:
```suggestion
	Context("when the server does not implement the config API", func() {
```

File: api/discovery/connector.go (hunk 2 above)
Comment (jmalloc) [reply api/discovery/connector.go]: Cheers, I've fixed the `obs` reference, that was an old parameter name, but the "It blah blah blah" is the standard way function documentation describes behavior in the standard libary.

File: api/discovery/connector.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+// watch checks if conn supports the config API and notifies the observer
+// accordingly.
+//
+// It blocks until ctx is canceled or the connection is severed, at which point
+// the observer is notified of the disconnection.
+func (w *watcher) watch(
+	ctx context.Context,
+	conn *grpc.ClientConn,
+) error {
+	// Make a context that closes the gRPC stream when this function exists, as
```
Comment (koden-km) [inline api/discovery/connector.go]:
```suggestion
	// Make a context that closes the gRPC stream when this function exits, as
```

File: api/discovery/doc.go [hunk 5]
```diff
@@ -0,0 +1,2 @@
+// Package discovery provides mechanisms for discovery config API servers.
```
Comment (koden-km) [inline api/discovery/doc.go]: Doc block is missing something here. Either "for _discovering_ config API servers" or "for discovery _of_ config API servers".
```suggestion
// Package discovery provides mechanisms for discovery of config API servers.
```

File: api/discovery/target.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+	// The options must not include grpc.WithBlock().
+	Options []grpc.DialOption
+
+	// MetaData contains driver-specific meta-data about the target.
+	MetaData MetaData
+}
+
+// TargetObserver is notified when config API targets are discovered.
+type TargetObserver interface {
+	// TargetAvailable is called when a target is becomes available.
```
Comment (koden-km) [inline api/discovery/target.go]:
```suggestion
	// TargetAvailable is called when a target becomes available.
```

**Pull request merged**